### PR TITLE
[ITS] Service and Support Article migrations

### LINKED ISSUE #7479: Migrate ITS services
# Story

As an ITS site owner, I would like our services content from our current site to be migrated to our new site so that we don't have to manually copy all of that content over.

# Additional Info

Migrate all published D7 infocenters to D10 services ~182

**Mappings - D10 - D7**

- Title - title
- Summary - body summary
- Description - filtered html - body
- Aliases - multiple - field_ic_aliases
- Audience - multiple - taxonomy - field_audience
  - terms already exist on new site
  - new site missing "Staff" term. Not sure if that was by design?
  - **ITS considering not having Audience taxonomy**
- Category - single - taxonomy - field_ic_category
  - terms already exist on new site
  - new site missing "Artificial Intelligence".
- Related Services - multiple - entity reference - field_related_info_centers
- Website - field_ic_website - URL and Title
- Quick Links - multiple - migrating from both of the following fields into this field
  - field_ic_obtain - URL and Title
  - field_ic_commonactions - URL and Title
- Contact - filtered html - field_ic_contactinfo
- Fees - field_ic_fee_info
  - No Charge or Other. If Other then there should be Additional Fee Information
- Additional Fee Information - field_ic_fees
- Authored On date

# Questions

Did the staff audience get intentionally left off the new site? - They are trying to determine if they want to keep the audience taxonomy at all.

Does the Artificial Intelligence category need to be added to new site? - Yes

Url Alias - no need to map and migrate custom url aliases that have been set for services.
Redirects - they do want us to create redirects from the previous url alias to the content on the new site.

# Proposed Solution

- Create an `its_migrate` module in the its.uiowa.edu site directory following other site migration patterns (configuration held in `module/config/split/*`, `.install` file to load config, etc)
- Create a Services YAML migration definition including all fields defined above _except_ Related Services
- Create a Services migrate source class that extends `sitenow_migrate::BaseNodeSource`
- Preprocess as necessary (eg Category term mapping; see `now_migrate::NewsFeature` for an example that creates tags on the fly; this can be adjusted to only look up current tags, as all tags should exist in the site already)
- Add a Services Redirects migration YAML definition
- Add a Post Migration Event Listener or a second Services YAML migration definition for the related services field
  - If an event listener, include a boolean flag somewhere (likely the `$migration` object) that indicates when the listener has run, and skip if bool is there to avoid second firings of the listener, a previous problem with migration listeners
  - If a second migration, follow the redirects migration pattern (utilizing the `Services` migration class, the `migration_lookup` process, etc), and adjust to map only the Related Services field. This will technically be an update migration, so slightly different than our previous migrations

### LINKED ISSUE #7519: Migrate ITS Support Articles
# Story

As an ITS site owner, I would like to have our support articles migrated from our D7 site to our new site so that we do not have to manually re-create each one.

# Additional Info

Migrate all published D7 support articles to D10 support articles ~1000

**Mappings - D10 - D7**

- Title - title
- Short Description - field_short_description
- Summary - body summary
- Body - filtered html - body
- Service - single - entity reference - field_info_center
- Category - single - reference - field_sa_category
- Faqs - multiple - field collection to paragraph - field_sa_faq
- Authored On date

Url Alias - no need to map and migrate custom url aliases that have been set for support articles.
Redirects - create redirects from the previous url alias to the content on the new site.

# Questions

Are they going to create support article category terms prior to migration or do we need to create on import if they do not exist?

# Proposed solution

```[tasklist]
### Tasks
- [x] Referencing recent migrations, attempt to use yaml mapping for most of this content migration, and string match lookups for the reference fields.
- [x] Reference and setup a redirect migration
- [ ] Given that there are 1000+ items, consider batching
- [ ] Consider establishing a paragraphs processing plugin if one does not already exist
```

### PR DESCRIPTION
**Pull request by GaryRidgway**
```
ddev blt ds --site its.uiowa.edu --sync-public-files && ddev drush @its.local en its_migrate -y && ddev drush @its.local uli /admin/config/sitenow/sitenow-migrate
```

- It is recommended that Service and Support article content is deleted before running.
- Warnings about FieldLink.php can be ignored. https://iowaweb.slack.com/archives/C0164FNN1PV/p1712763861770599?thread_ts=1712759044.122119&cid=C0164FNN1PV

## Services

Resolves https://github.com/uiowa/uiowa/issues/7479

- Run the migrations for service, and then the service redirects
```
ddev drush @its.local mim its_service && ddev drush @its.local mim its_service_redirects
```

- Note that there are broken links for services, and that they will output in the console when you run the migrations
- Also note that there will be a few Services that are created because of broken service connections on previously populated Info center content types. Delete these when you log in to check after the migration, as that is what will happen after we run it.
- Find your favorite service from d7, and ensure that its fields are mapped as shown in the checklist below. You may have to look at multiple services to find all the fields. The 'Wiki' service is a good starting point.
- Check that related services is populated
- Now look at the redirects from d7
- Check that those have been moved over. 
<details><summary>Checklist for services</summary>
<p>

- [x] Aliases - multiple - field_ic_aliases
- [x] Audience - multiple - taxonomy - field_audience
    - [x] terms already exist on new site
    - [x] new site missing "Staff" term. Not sure if that was by design?
    - [x] **ITS considering not having Audience taxonomy**
- [x] Category - single - taxonomy - field_ic_category
    - [x] terms already exist on new site
    - [x] new site missing "Artificial Intelligence".
- [x] Website - field_ic_website - URL and Title
- [x] Quick Links - multiple - migrating from both of the following fields into this field
    - [x] field_ic_obtain - URL and Title
    - [x] field_ic_commonactions - URL and Title
- [x] Contact - filtered html - field_ic_contactinfo
- [x] Fees - field_ic_fee_info
    - [x] No Charge or Other. If Other then there should be Additional Fee Information
- [x] Additional Fee Information - field_ic_fees
- [x] Related Services - multiple - entity reference - field_related_info_centers
- [x] Description - filtered html - body
- [x] Summary - body summary
- [x] Title - title
- [x] Authored On date

</p>
</details>

## Support articles

Resolves https://github.com/uiowa/uiowa/issues/7519

_Assumes `ddev drush @its.local mim its_service` has already run_
```
ddev drush @its.local mim its_support_articles --limit=100
```

- Click through a couple support articles. All of the fields should be accounted for. Title/Body/Date
- FAQs should be present and in order on `Connecting a mobile device to the eduroam network` with media.
- Short description should be on some nodes like `Adding Social Media Content`
- Category and service should be on almost all support articles. The service should be linked in the breadcrumb and the service page should show the support articles in the view.
- It attempts to fix some broken links.

```
ddev drush @its.local mim its_support_articles_redirects
```

- Test redirects by recalling your favorite support article ids (e.g. support/article/283) and see that it takes you to the right spot.

### CODE REVIEW

**Comment by cory-skeers** [comment]
Updated to include audience and category taxonomies.

- Can be removed if they decide to not continue with an audience taxonomy
- It's set to look for only terms that exist, and not create new. Anything it doesn't find it includes a message in the migration table. But this could be updated to create new terms if they don't already exist
- For slight simplicity, it looks for terms by name and assumes that there is no overlap between vocabularies (eg there's not a duplicate term in both Service Category and Support Article taxonomies)
  - Update: Added a todo here, because it looks like there are some overlaps...

**Comment by GaryRidgway** [comment]
From my conversation with @joewhitsitt, I think this is now reviewable? We just need to keep in mind that afterwards we are going to want to delete the content created automatically with the migration.

File: docroot/sites/its.uiowa.edu/modules/its_migrate/src/Plugin/migrate/source/Service.php [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+  /**
+   * Helper function to fetch existing tags.
+   */
+  private function fetchTag($tag_name, $row) {
+    // Check if we already have the tag in the destination.
+    $result = \Drupal::database()
+      ->select('taxonomy_term_field_data', 't')
+      ->fields('t', ['tid'])
+      // @todo Add another conditional to match the proper vocabulary,
```
Comment (joewhitsitt) [inline docroot/sites/its.uiowa.edu/modules/its_migrate/src/Plugin/migrate/source/Service.php]: Left over todo

Comment (GaryRidgway) [reply docroot/sites/its.uiowa.edu/modules/its_migrate/src/Plugin/migrate/source/Service.php]: working on this

**Comment by GaryRidgway** [comment]
I was checking the additions I made for the link breakage reporting, but I was receiving no more errors than normal, but @joewhitsitt reported that he was seeing some when doing the services migration. I think I am going to try and do the whole migration all the way through with `dblog` and seeing if it passes with clean databases from both d7 and d10.

**Comment by cory-skeers** [comment]
For services, I'm Noticing that we're not capturing the signup info anywhere

image

In the d7 site it's a separate field. Do we need to add a field to the new site, or append it an existing field like the body or contact? There are 32 services that have info in this field with various types of details and/or contact info.

**Comment by cory-skeers** [comment]
Some (63) of the quick links in services are absolute URLs to its.uiowa.edu pages. Most of these are to support articles, with a small handful going other places. We could do another sub-process to update these based on the support article migration map, flag/report them, and/or leave them in place if the its folks are planning on doing a thorough check and edit with these.

**Comment by joewhitsitt** [comment]
>Noticing that we're not capturing the signup info anywhere

I don't see the signup field present on https://github.com/uiowa/uiowa/issues/6666 or https://github.com/uiowa/uiowa/issues/6806.

Do you recall @briand44

**Comment by joewhitsitt** [comment]
>We could do another sub-process to update these based on the support article migration map, flag/report them, and/or leave them in place if the its folks are planning on doing a thorough check and edit with these.

Part of why we didn't do broken links checks for the services is because it affected the existing sub_process to map existing services. It ended up creating duplicates/exiting early. So service body broken links aren't even reported. Alan previously mentioned that linkchecker has been on the ITS website for years and many services have broken links reported already. Not sure it is worth doing anymore than mentioning at this point.

**Comment by cory-skeers** [comment]
> > Noticing that we're not capturing the signup info anywhere
> 
> I don't see the signup field present on #6666 or #6806.
> 
> Do you recall @briand44

There's also a Getting Help field, that from #6806 I'm thinking should also be put into the Contact field (appended following anything else from the Contact Info field)?

**Comment by cory-skeers** [comment]
I think other than checking on the signup and getting help fields and whether to / where to map those, testing on these has all checked out 💯

**Comment by briand44** [comment]
Is there a way to provide a list of the InfoCenters that have Sign Up and Getting Help info?


> We're planning on moving the content in those fields (or in a lot of cases just deleting it). Having both the Getting help and Contact fields are redundant, so we're keeping the contact one and pitching the other. The sign up field, was also causing some confusion and was intended for people to literally sign up for access to software, but editors weren't using it as intended. 
> 
> Is there any way to capture the info centers that have those fields filled out so that we can move the content before migration? Can we put it in an excel spreadsheet or something?

**Comment by joewhitsitt** [comment]
views_data_export is currently turned on in ITS d7 prod. we could get a quick list/output from there? is that the kind of thing you are looking for @briand44 or do you want it tied to future website nodes?

**Comment by briand44** [comment]
>is that the kind of thing you are looking for

Yes, I think they would be happy with that.

Is that something we can provide as part of this work or would you like me to create a separate issue for that?

**Comment by cory-skeers** [comment]
> Is there a way to provide a list of the InfoCenters that have Sign Up and Getting Help info?

@briand44 Sure thing--node ids and titles below.

<details>
<summary>
Info centers with "Getting Help" content
</summary>

| NID | Title |
| ------------------ | ----------------- |
|1771|Qualtrics|
|100003|Project Management Office|
|100129|LAMP|
|100144|High Performance Computing|
|100177|Galaxy|
|100179|SBGrid|
|100181|REDCap (Research Electric Data Capture)|
|100284|Cable TV|
|100293|Vulnerability Scanning & Penetration Testing|
|100295|TLS Certificates for Secure Communications|
|100297|Security Consulting, Reviews and Assessments|
|100306|Azure Dev Tools for Teaching|
|100317|SITA - Student Instructional Technology Assistants|
|100323|Video Surveillance|
|100325|Accessibility and Assistive Technology|
|100333|Account/Identity Management|
|100337|Campus Data Network Connectivity|
|100349|Computer Data and Media Disposal Policy Training|
|100393|Telephone and Voice Services|
|100564|Innovations in Teaching with Technology Awards (ITTA)|
|102190|RemoteAdmin Service for Access to Secured Servers and Administrative Applications|
|103597|FileVault Encryption|
|103906|Large Scale Storage (LSS)|
|105606|Amazon Web Services (AWS)|
|105611|Microsoft Azure|
|108546|MatLab|
|115766|Interactive Data Analytics Service (IDAS)|
|117101|Microsoft Teams|
|118236|Microsoft Planner|
|119836|Core Application Services|
|122331|CloudCheckr|
|122706|Password Vault|
|125121|Iowa Health Data Resource Data Enclave|

</details>


<details>
<summary>
Info centers with "Signup" content
</summary>

| NID | Title |
| ------------------ | ----------------- |
|205|Wiki|
|261|Printing in Computer Labs (ITCs)|
|284|MyWeb|
|2050|Plagiarism Detection|
|100049|Microsoft Configuration Manager (MCM)|
|100144|High Performance Computing|
|100177|Galaxy|
|100180|Research Data Collaboration Service (RDCS)|
|100313|Available Software|
|100323|Video Surveillance|
|100329|Tomcat|
|100332|Custom Application Development & Integration|
|100346|Hawk Alert|
|100347|A/V Design & Installation|
|100352|Web Hosting (Windows/IIS)|
|101234|UI-Guest Wi-Fi|
|102155|JAMF Pro Apple Device Management|
|102577|Bitbucket Server|
|105606|Amazon Web Services (AWS)|
|105611|Microsoft Azure|
|109041|ICON Direct|
|109251|UI Learn|
|110071|Research Remote Desktop Service|
|114521|GitLab Administrative|
|115171|Puppet|
|119401|Bookings|
|119846|Managed Application and Platform Level Services|
|122191|Office 365 GCC High|
|122331|CloudCheckr|
|122706|Password Vault|
|123541|Linux Software Mirrors|
|123901|Windows Encryption (Bitlocker)|

</details>

**Comment by cory-skeers** [comment]
> > is that the kind of thing you are looking for
> 
> Yes, I think they would be happy with that.
> 
> Is that something we can provide as part of this work or would you like me to create a separate issue for that?

Ope. Was already pulling stuff out in https://github.com/uiowa/uiowa/pull/7524#issuecomment-2057689449, grabbed it from the local copy I synced down today.

**Review by cory-skeers — APPROVED** [review]
I think along with https://github.com/uiowa/uiowa/pull/7524#issuecomment-2057689449 to provide them with, this is good to go! 🚀

**Pull request merged**